code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/ManualCarouselNotificationBuilderTest.kt

```diff
@@ -0,0 +1,239 @@
+            context,
+            pushTemplate,
+            trackerActivityClass,
+            broadcastReceiverClass
+        )
+        verify(exactly = 0) {
+            BasicNotificationBuilder.fallbackToBasicNotification(
+                context,
+                trackerActivityClass,
+                broadcastReceiverClass,
+                pushTemplate.data
+            )
+        }
+    }
+
+    @Test
+    fun `test downloadCarouselItems return non empty list for it retrieve the image bitmap from cache successfully`() {
+        every { getCachedImage(any()) } answers { mockkClass(Bitmap::class) }
+        val imagesList =
+            ManualCarouselNotificationBuilder.downloadCarouselItems(pushTemplate.carouselItems)
+        assertFalse(imagesList.isEmpty())
+    }
+
+    @Test
+    fun `test getCarouselIndices with left click intent action`() {
+        pushTemplate.intentAction = PushTemplateConstants.IntentActions.MANUAL_CAROUSEL_LEFT_CLICKED
+        val imageUris = listOf("image1", "image2", "image3")
+        val result = getCarouselIndices(pushTemplate, imageUris)
+        assertEquals(Triple(1, 2, 0), result)
+    }
+
+    @Test
+    fun `test getCarouselIndices with filmstrip left click intent action`() {
+        pushTemplate.intentAction = PushTemplateConstants.IntentActions.FILMSTRIP_LEFT_CLICKED
+        val imageUris = listOf("image1", "image2", "image3")
+        val result = getCarouselIndices(pushTemplate, imageUris)
+        assertEquals(Triple(1, 2, 0), result)
+    }
+
+    @Test
+    fun `test getCarouselIndices with no intent action and filmstrip layout`() {
+        pushTemplate.intentAction = PushTemplateConstants.DefaultValues.FILMSTRIP_CAROUSEL_MODE
+        val imageUris = listOf("image1", "image2", "image3")
+        val result = getCarouselIndices(pushTemplate, imageUris)
+        assertEquals(
+            Triple(
+                PushTemplateConstants.DefaultValues.FILMSTRIP_CAROUSEL_CENTER_INDEX - 1,
+                PushTemplateConstants.DefaultValues.FILMSTRIP_CAROUSEL_CENTER_INDEX,
+                PushTemplateConstants.DefaultValues.FILMSTRIP_CAROUSEL_CENTER_INDEX + 1
+            ),
+            result
+        )
+    }
+
+    @Test
+    fun `test populateManualCarouselImages with valid images`() {
+        val packageName = context.packageName
+        val centerIndex = 1
+        every { getCachedImage(any()) } answers { mockkClass(Bitmap::class) }
+        every { expandedLayout.addView(any(), any<RemoteViews>()) } just Runs
+        every { expandedLayout.setDisplayedChild(any(), any()) } just Runs
+
+        populateManualCarouselImages(
+            context,
+            pushTemplate.carouselItems,
+            packageName,
+            centerIndex,
+            pushTemplate,
+            trackerActivityClass,
+            expandedLayout
+        )
+
+        verify { expandedLayout.addView(any(), any<RemoteViews>()) }
+        verify { expandedLayout.setDisplayedChild(any(), centerIndex) }
```

Comment:
Is it also possible to verify the correct image and caption is set to the RemoteView by spying and verifying that `setImageViewBitmap` and `setTextViewText` are called with the expected values?

Reply:
One more test to be added if possible is for `setRemoteViewClickAction` to make sure the click action has the right uri. Verify if uri is missing for the carousel item, then the template's `actionUri` is used

Reply:
done @spoorthipujariadobe